feat(doc): include bao test framework introduction (#71)
bao-project/bao-docs: source/development/testing.rst

Testing
=======

> The primary goal of the Bao Test Framework is to optimize the testing process
> of Bao. It encompasses three core components: (i) a C code library designed for

Reviewer: ```suggestion
The primary goal of the Bao Test Framework is to provide the infrastructure for the testing process
of Bao Project components. It encompasses three core components: (i) a C code library designed for
```

Author: Done!

real-time test handling, (ii) a build system based on Nix, and (iii) a Python
API for comprehensive test management.

> 1. **C Code Library** - This library comprises a set of utilities that simplify

Reviewer: ```suggestion
1. **C Library** - This library comprises a set of utilities that simplify
```

C library is enough, "C code" is kind of a pleonasm.

Author: Agree, integrated.

   test definition and facilitate result assessment, including streamlined
   result logging;

2. **Nix Build System** - The Nix Build System introduces an abstraction layer
   for system builds, enabling the construction of various components, from the
   firmware to the guests running in each test;

> 3. **Python API** - The Python API adds a layer of abstraction to test

Reviewer: Do we really want to call this an API? Isn't this actually a tool that is invoked through the command line?

Author: You're correct, updated to "Python Tool".

Testing
=======
The primary goal of the Bao Test Framework is to provide the infrastructure for
the testing process of Bao Project components. It encompasses three core

> components: (i) a C code library designed for real-time test handling, (ii) a

Reviewer 2: ```suggestion
components: (i) a C library designed for real-time test handling, (ii) a
```

Author: Thanks, fixed!

build system based on Nix, and (iii) a Python Tool for comprehensive test
management.

1. **C Library** - This library comprises a set of utilities that simplify test
   definition and facilitate result assessment, including streamlined result
   logging;

2. **Nix Build System** - The Nix Build System introduces an abstraction layer
   for system builds, enabling the construction of various components, from the
   firmware to the guests running in each test;

> 3. **Python Tool** - The Python Tool adds a layer of abstraction to test
>    deployment, managing both the Nix Build System and the C code library
>    components.

Reviewer: I'd like a bit more of detail. "layer of abstraction" is too abstract. I'd suggest something in the lines of :+1: 
```suggestion
3. **Python Tool** - A command line tool that drives the framework by 
 managing both the Nix Build System and the C code library components
 to build testing artifacts, automating their deployment in the target platform,
 and gathering, analyzing and outputting the final results.
```

Author: Agree, thanks! Fixed.

Testing
=======
The primary goal of the Bao Test Framework is to provide the infrastructure for the testing process
of Bao Project components. It encompasses three core components: (i) a C library designed for
real-time test handling, (ii) a build system based on Nix, and (iii) a Python Tool for
comprehensive test management.

> 1. **C Library** - This library that streamlines testing by providing macros, pre-processed logging
>    prompts, and an entry-point-based system, aiming to facilitating the integration of developer
>    tests in the software stack, such as the hypervisor or guest components.

Reviewer: ```suggestion
1. **C Library** - This library provides a framework for writing tests. It defines (i) a runtime that automates the invocation of function unit tests and (ii) an API that mainly provides a way of outputing test results in a standard format that can be parsed by the Python Tool (see below).
```

Author: Thanks, fixed!

2. **Nix Build System** - The Nix Build System introduces an abstraction layer for system builds,
   facilitating a modular construction of the software stack for comprehensive testing. It

>    encompasses a series of nix recipes allow the compilation and build of three distinct layers:

Reviewer 2: ```suggestion
   encompasses a series of nix recipes that allow the compilation and build of three distinct layers:
```

> 2. **Nix Build System** - The Nix Build System introduces an abstraction layer for system builds,
>    facilitating a modular construction of the software stack for comprehensive testing. It
>    encompasses a series of nix recipes that allow the compilation and build of three distinct
>    layers: (i) the hypervisor, (ii) multiple guests (e.g., baremetal, freeRTOS, etc.), and (iii)
>    the firmware.

Reviewer: ```suggestion
2. **Nix Build System** - The Nix Build System introduces an abstraction layer for system builds,
   facilitating a modular construction of the target software stack used for testing. It
   encompasses a series of nix recipes that allow the compilation and build of three distinct
   layers: (i) the hypervisor, (ii) multiple guests (e.g., baremetal, freeRTOS, etc.), and (iii)
   the firmware.
```

Author: Thanks, fixed!

Testing
=======
The primary goal of the Bao Test Framework is to provide the infrastructure for
the testing process of Bao Project components. It encompasses three core
components: (i) a C library designed for real-time test handling, (ii) a
build system based on Nix, and (iii) a Python Tool for comprehensive test
management.

> 1. **C Library** - This library comprises a set of utilities to streamline test
>    definitions by introducing specialized testing function macros and
>    simplifying result assessment through pre-processed logging prompts.
>    Additionally, it incorporates an entry-point-based system, fostering seamless
>    integration between developer tests and the software stack, such as the
>    hypervisor or guest components.

Reviewer 2: I'd simplify the language  a little bit. 
```suggestion
1. **C Library** - This library that streamlines testing by providing macros,
   pre-processed logging prompts, and an entry-point-based system, aiming to
   facilitating the integration of developer tests in the software stack, such
   as the hypervisor or guest components.
```

Author: Fixed, thanks!

2. **Nix Build System** - The Nix Build System introduces an abstraction layer
   for system builds, facilitating a modular construction of the software stack

>    for comprehensive testing. It encompasses a series of nix recipes
>    meticulously crafted to enable the compilation and build of three distinct

Reviewer 2: ```suggestion
   for comprehensive testing. It encompasses a series of nix recipes
   allow the compilation and build of three distinct
```

Author: Fixed, thanks!
